src/hil/spi_master.rs

```diff
@@ -1,73 +1,86 @@
-//! Traits and parameters for SPI master communication
+//! Traits and parameters for SPI master communication.
+//! Authors: Sam Crow <samcrow@uw.edu>
+//!          Philip Levis <pal@cs.stanfored.edu>
 
 use core::option::Option;
 
-#[derive(Copy, Clone)]
-pub enum Rate {
-    MSBFirst,
-    LSBFirst,
-}
 /// Values for the ordering of bits
 #[derive(Copy, Clone)]
-pub enum DataOrder {
-    MSBFirst,
-    LSBFirst,
-}
+pub enum DataOrder {MSBFirst, LSBFirst}
 
 /// Values for the clock polarity (idle state or CPOL)
 #[derive(Copy, Clone)]
-pub enum ClockPolarity {
-    IdleHigh,
-    IdleLow,
-}
-/// Values for the clock phase (CPHA), which defines when
-/// values are sampled
+pub enum ClockPolarity {IdleHigh, IdleLow}
+
+/// Which clock edge values are sampled on
 #[derive(Copy, Clone)]
-pub enum ClockPhase {
-    SampleLeading,
-    SampleTrailing,
-}
+pub enum ClockPhase {SampleLeading, SampleTrailing}
 
 pub trait SpiCallback {
     /// Called when a read/write operation finishes
-    fn read_write_done(&'static self, 
-                       read: Option<&'static mut[u8]>,
-                       writer: Option<&'static mut[u8]>);
```

Comment:
OK, I don't understand the second option then. How is the driver getting access to the in-kernel buffer each time? Doesn't it that buffer have to be passed up from the callback? The driver is probably internally storing it inside an `Option` (or something), then `take()`ing from the option and passing it to the HAL. But it needs a way to get a hold of the buffer again -- like, getting it back from the callback.

Again, you're not running into this problem yet because your client code uses `unsafe` so you can just reference a mutable static directly. But I think you'll run into problems implementing a driver.

Happy to cross that bridge when we get to it though